File: 01-backend-node-native/src/domain/services/auth.service.ts
import { ServerResponse } from 'http';
import { RepositoryFactory } from '@infra/repository.factory';
import { AuthUtil } from '@shared/utils/auth.util';
import { CreateUserDto } from '@domain-types/user.types';
import { BadRequestError, UnauthorizedError } from '@shared/errors/app.error';
import { AuthFactory } from '@api/middleware/auth/auth.factory';
import { logger } from '@shared/utils/logger';

const userRepo = RepositoryFactory.getUserRepository();

export class AuthService {
  static async register(data: CreateUserDto, res: ServerResponse) {
    const existingUser = await userRepo.findByEmail(data.email);
    if (existingUser) {
      logger.warn(`⚠️ Registration failed: Email ${data.email} already exists`);
      throw new BadRequestError('User already exists');
    }

    const hashedPassword = await AuthUtil.hashPassword(data.password!);
    
    // Industrial logic: First user is automatically Admin
    const allUsers = await userRepo.findAll();
    const role = allUsers.length === 0 ? 'admin' : (data.role || 'user');

    const user = await userRepo.create({ ...data, password: hashedPassword, role });
    
    logger.info(`👤 New User registered: ${user.email} (ID: ${user.id}) as ${user.role}`);

    // Use Strategy to handle success (Token or Cookie)
    const authData = AuthFactory.getStrategy().onAuthSuccess(res, user);
    
    return { user, ...authData };
  }

  static async login(email: string, password: string, res: ServerResponse) {
    const user = await userRepo.findByEmail(email);
    if (!user) {
      logger.warn(`❌ Login attempt failed: User not found (${email})`);
      throw new UnauthorizedError('Invalid credentials');
    }

    const isValid = await AuthUtil.comparePassword(password, user.password!);
    if (!isValid) {
      logger.warn(`❌ Login attempt failed: Incorrect password for ${email}`);
      throw new UnauthorizedError('Invalid credentials');
    }

    // Use Strategy to handle success (Token or Cookie)
    const authData = AuthFactory.getStrategy().onAuthSuccess(res, user);
    
    logger.info(`✅ User logged in successfully: ${user.email}`);

    return { user, ...authData };
  }
}

File: 04-backend-springboot/src/main/java/com/smarttask/ai/domain/services/AuthService.java
package com.smarttask.ai.domain.services;

import com.smarttask.ai.domain.User;
import com.smarttask.ai.infrastructure.repositories.UserRepository;
import com.smarttask.ai.shared.utils.HashUtil;
import com.smarttask.ai.shared.utils.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    public Map<String, Object> register(String name, String email, String password) {
        Optional<User> existingUser = userRepository.findByEmail(email);
        if (existingUser.isPresent()) {
            throw new RuntimeException("User already exists");
        }

        String hashedPassword = HashUtil.hashPassword(password);
        
        // First user is Admin
        long count = userRepository.count();
        String role = count == 0 ? "admin" : "user";

        User user = User.builder()
                .name(name)
                .email(email)
                .password(hashedPassword)
                .role(role)
                .build();

        user = userRepository.save(user);

        Map<String, Object> result = new HashMap<>();
        result.put("user", user);

        String authStrategy = System.getProperty("AUTH_TYPE", "JWT");
        if ("JWT".equalsIgnoreCase(authStrategy)) {
            String token = JwtUtil.generateToken(user.getId(), user.getEmail());
            result.put("token", token);
        }

        return result;
    }

    public Map<String, Object> login(String email, String password) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Invalid credentials"));

        if (!HashUtil.comparePassword(password, user.getPassword())) {
            throw new RuntimeException("Invalid credentials");
        }

        Map<String, Object> result = new HashMap<>();
        result.put("user", user);

        String authStrategy = System.getProperty("AUTH_TYPE", "JWT");
        if ("JWT".equalsIgnoreCase(authStrategy)) {
            String token = JwtUtil.generateToken(user.getId(), user.getEmail());
            result.put("token", token);
        }

        return result;
    }
}
